nntrainer/utils/profiler.h

```diff
@@ -0,0 +1,254 @@
+ *
+ * @return std::string name
+ */
+std::string event_to_str(const int event);
+
+class Profiler;
+/**
+ * @brief Generic profile listener class to attach to a profiler,
+ * this can be inherited to create a custom profile listener
+ */
+class ProfileListener {
+public:
+  /**
+   * @brief Construct a new Profile Listener object
+   *
+   * @param profiler_ profiler that this listener is bound to. Unsubscribe will
+   * be called when destruction
+   * @param events events for this profiler to listen to
+   */
+  ProfileListener(Profiler *profiler_, std::vector<int> events);
+
+  /**
+   * @brief Destroy the Base Profile Listener object
+   *
+   */
+  virtual ~ProfileListener() noexcept;
+
+  /**
+   * @brief A callback function to be called from a profiler
+   *
+   * @param event event key to store the result
+   * @param value time value from the profiler
+   */
+  virtual void onNotify(const int event,
+                        const std::chrono::milliseconds &value) = 0;
+
+  /**
+   * @brief resets the listener to the inital state for a particular key
+   *
+   * @param event event which profiler should notice
+   */
+  virtual void reset(const int event) = 0;
+
+  /**
+   * @brief get the latest result of a event
+   *
+   * @param event event to query the result
+   * @return const std::chrono::milliseconds
+   */
+  virtual const std::chrono::milliseconds result(const int event) = 0;
+
+  /**
+   * @brief report the result
+   *
+   * @param out outstream object to make a report
+   */
+  virtual void report(std::ostream &out) const = 0;
+
+private:
+  Profiler *profiler;
+};
+
+class GenericProfileListener : public ProfileListener {
+public:
+  /**
+   * @brief Construct a new GenericProfile Listener object
+   *
+   * @param profiler profiler that this listener is bound to. pass null if empty
+   * @param warmups_ ignore first @a warmups_ records when making report
+   */
+  GenericProfileListener(Profiler *profiler, std::vector<int> events = {},
+                         int warmups_ = 0) :
+    ProfileListener(profiler, events),
+    warmups(warmups_) {}
+
+  /**
+   * @brief Destroy the Generic Profile Listener object
+   *
+   */
+  virtual ~GenericProfileListener() = default;
+
+  /**
+   * @copydoc ProfileListener::onNotify(const int event, const
+   * std::chrono::milliseconds &value)
+   */
+  virtual void onNotify(const int event,
+                        const std::chrono::milliseconds &value) override;
+
+  /**
+   * @copydoc ProfileListener::reset(const int event)
+   */
+  virtual void reset(const int event) override;
+
+  /**
+   * @copydoc ProfileListener::result(const int event)
+   */
+  virtual const std::chrono::milliseconds result(const int event) override;
+
+  /**
+   * @copydoc ProfileListener::report(std::ostream &out)
+   */
+  virtual void report(std::ostream &out) const override;
+
+private:
+  unsigned int warmups;
+
+  std::unordered_map<int, std::vector<std::chrono::milliseconds>> time_taken;
```

Comment:
I encourage you to consider using sum of all times and the count rather than this vector. This will grow big very fast.
You can add more statistics than just sum, but vector in a profiler will slow it down.

Reply:
Done